pkg/provider/oidc.go

```diff
@@ -0,0 +1,195 @@
+	ErrInvalidToken         = errors.New("invalid token")
+	ErrInvalidAuthorization = errors.New("invalid authorization")
+)
+
+func newRelayingPartyOIDC(ctx context.Context, oidc *configurationOIDC) (rp.RelyingParty, error) {
+	redirectURI := fmt.Sprintf("http://localhost:%v%v", oidc.port, oidc.callbackURL)
+	cookieHandler := httphelper.NewCookieHandler(key, key, httphelper.WithUnsecure())
+	options := []rp.Option{
+		rp.WithCookieHandler(cookieHandler),
+		rp.WithVerifierOpts(rp.WithIssuedAtOffset(5 * time.Second)),
+	}
+	if oidc.clientSecret == "" {
+		options = append(options, rp.WithPKCE(cookieHandler))
+	}
+	if oidc.keyPath != "" {
+		options = append(options, rp.WithJWTProfile(rp.SignerFromKeyPath(oidc.keyPath)))
+	}
+	return rp.NewRelyingPartyOIDC(ctx, oidc.issuer, oidc.clientID, oidc.clientSecret, redirectURI, oidc.scopes, options...)
+}
+
+func marshalUserinfo() rp.CodeExchangeUserinfoCallback[*oidc.IDTokenClaims, *oidc.UserInfo] {
+	return func(w http.ResponseWriter, r *http.Request, tokens *oidc.Tokens[*oidc.IDTokenClaims], state string, rp rp.RelyingParty, info *oidc.UserInfo) {
+		data, err := json.Marshal(info)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		w.Write(data)
+	}
+}
+
+func oidcAuthenticatedHTTPInterceptor(resourceServer rs.ResourceServer) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			ctx := req.Context()
+
+			token, err := checkOIDCTokenFromRequest(req)
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusUnauthorized)
+				return
+			}
+
+			pi := info.ProviderInfoFromContext(ctx)
+			resp, err := checkIntrospectWithCached(ctx, resourceServer, token, pi.IntrospectionResponse)
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusForbidden)
+				return
+			}
+			ctx = pi.SetIntrospectionResponse(resp).IntoContext(ctx)
+
+			next.ServeHTTP(w, req.WithContext(ctx))
+		})
+	}
+}
+func oidcAuthorizedHTTPInterceptor(resourceServer rs.ResourceServer, requestedClaim, requestedValue string) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			ctx := req.Context()
+
+			token, err := checkOIDCTokenFromRequest(req)
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusUnauthorized)
+				return
+			}
+
+			pi := info.ProviderInfoFromContext(ctx)
+			resp, err := checkIntrospectWithCached(ctx, resourceServer, token, pi.IntrospectionResponse)
+			if err != nil {
+				http.Error(w, err.Error(), http.StatusForbidden)
+				return
+			}
+			ctx = pi.SetIntrospectionResponse(resp).IntoContext(ctx)
+
+			if err := checkAuthorization(pi.IntrospectionResponse, requestedClaim, requestedValue); err != nil {
+				http.Error(w, err.Error(), http.StatusForbidden)
+				return
+			}
+			next.ServeHTTP(w, req.WithContext(ctx))
```

Comment:
mostly redundant (apart from the `checkAuthorization`) and why can on string claims be checked, resp. shouldn't there be multiple options what to be checked (simple authZ check, role check, custom checks, ...)

Reply:
can be expanded, started of with this as an example